src/compiler/parser.ts

```diff
@@ -6601,6 +6611,25 @@ namespace Parser {
             case SyntaxKind.NewKeyword:
                 return parseNewExpressionOrNewDotTarget();
             case SyntaxKind.SlashToken:
+                // JSX inside a .ts (not .tsx) file gets a specialized error...

```

Comment:
From https://github.com/microsoft/TypeScript/pull/56101#pullrequestreview-1735462101:

> The change in parsePrimaryExpression is pretty invasive whereas the parseTypeAssertion is simple. How much of the benefit comes from the parseTypeAssertion change?

Hard to say what balance comes from parsing self-closing tags (`parseTypeAssertion`) vs. tags with children (`parsePrimaryExpression`). My _intuition_ is that tags with children (the one we find more invasive) is _probably_ much more common - especially in complex pages where the error messages without this change are convoluted. 😕